zcncore/zcnfmt.go

```diff
@@ -0,0 +1,121 @@
+package zcncore
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+const (
+	SAS Unit = iota
+	UZCN
+	MZCN
+	ZCN
+)
+
+var (
+	ErrUndefinedFormat = errors.New("undefined format")
+
+	// regexp to parse zcn
+	reParseZCN = regexp.MustCompile(`(\d+)\s+(SAS|sas|uZCN|uzcn|mZCN|mzcn|ZCN|zcn)`)
+)
+
+type Unit byte
+
+func (unit Unit) String() string {
+	switch unit {
+	case SAS:
+		return "SAS"
+	case MZCN:
+		return "mZCN"
+	case UZCN:
+		return "uZCN"
+	}
+	return "ZCN"
+}
+
+func (unit *Unit) Parse(s string) error {
+	switch s {
+	case "SAS", "sas":
+		*unit = SAS
+	case "uZCN", "uzcn":
+		*unit = UZCN
+	case "mZCN", "mzcn":
+		*unit = MZCN
+	case "ZCN", "zcn":
+		*unit = ZCN
+	default:
+		return ErrUndefinedFormat
+	}
+	return nil
+}
+
+func ParseUnit(s string) (Unit, error) {
+	var unit Unit
+	err := unit.Parse(s)
+	if err != nil {
+		return unit, err
+	}
+	return unit, nil
+}
+
+// Balance is a type used for 0chain native token
+type Balance int64
+
+func (token Balance) Format(unit Unit) string {
+	switch unit {
+	case SAS:
+		return fmt.Sprintf("%d SAS", token)
+	case UZCN:
+		return fmt.Sprintf("%.3f uZCN", float64(token)/1e3)
+	case MZCN:
+		return fmt.Sprintf("%.3f mZCN", float64(token)/1e6)
+	}
+	return fmt.Sprintf("%.3f ZCN", float64(token)/1e9)
```

Comment:
According to this issue [here](https://github.com/0chain/zwalletcli/issues/10#issue-891467717)
```
0.0000000001 ZCN = 0.0000001 mZCN = 0.0001 µZCN = 1 sas 
```
reverse of this 
```
1 µZCN = 1e4 sas
1 mZCN = 1e7 sas
1 ZCN = 1e10 sas
```
But i see you have considered only `1e9`

Reply:
Indeed the testFormat() function is wrong. I think the confusion lies is Saswatas not being a 3rd power of decimals like usual thousand separators. It should use 1e4, 1e7 and 1e10. Saswatas shouldn't really be displayed if we set threshold for displaying uzcn to 4 decimals, e.g. 1 sas = 0.0001 µZCN

Reply:
Just to clarify, the format chosen by autoformat is a separate layer to the amount of raw zcn being shown.

That is, there could be user preference to show amounts as zcn down to 0.01 ZCN, then below that e.g. 0.0099 would be shown as 9.9mZCN.

Reply:
@NoSkillGuy, thanks for pointing it out. I"ll fix the order as 1 zcn = 1e10 sas.